src/smolagents/models.py

```diff
@@ -1161,6 +1170,136 @@ def create_client(self):
+        self.custom_role_conversions = custom_role_conversions
+
+        super().__init__(
+            model_id=model_id,
+            custom_role_conversions=custom_role_conversions,
+            flatten_messages_as_text=flatten_messages_as_text,
+            **kwargs,
+        )
+
+    def create_client(self):
+        try:
+            import boto3
+        except ModuleNotFoundError as e:
+            raise ModuleNotFoundError(
+                "Please install 'aws_sdk' extra to use AmazonBedrockServerModel: `pip install 'smolagents[aws_sdk]'`"
+            ) from e
+
+        return boto3.client("bedrock-runtime")
+
+    def __call__(
+        self,
+        messages: List[Dict[str, str]],
+        tools_to_call_from: Optional[List[Tool]] = None,
+        **kwargs,
+    ) -> ChatMessage:
+        # smolagents inject automatic parameters like end_stop, Observation and others
+        # The Bedrock API only supports the fields explicitly defined here.
+        # To ensure compliance, `kwargs` is set to empty, and customers must provide additional parameters using `bedrock_api_kwargs`.
+        kwargs = {}
+
+        # Bedrock only supports `assistant` and `user` roles.
+        # Many Bedrock models do not allow conversations to start with the `assistant` role, so the default is set to `user/user`.
+        # This parameter is retained for future model implementations and extended support.
+        if not self.custom_role_conversions:
+            self.custom_role_conversions = {
+                MessageRole.SYSTEM: MessageRole.USER,
+                MessageRole.ASSISTANT: MessageRole.USER,
+                MessageRole.TOOL_CALL: MessageRole.USER,
+                MessageRole.TOOL_RESPONSE: MessageRole.USER,
+            }
+
+        completion_kwargs: Dict = self._prepare_completion_kwargs(
+            messages=messages,
+            tools_to_call_from=tools_to_call_from,
+            custom_role_conversions=self.custom_role_conversions,
+            convert_images_to_image_urls=True,
+            is_bedrock_runtime=True,
+            **kwargs,
+        )
+
+        # Preparing the request payload to bedrock-runtime API
+        request_payload: dict[str, Any] = {
+            "modelId": self.model_id,
+            **completion_kwargs,
+        }
+
+        if self.guardrail_config:
+            request_payload["guardrailConfig"] = self.guardrail_config
+
+        if self.inference_config:
+            request_payload["inferenceConfig"] = self.inference_config
+
+        if self.bedrock_api_kwargs:
+            if self.bedrock_api_kwargs.get("toolConfig"):
+                # Not all models in Bedrock support `toolConfig`. Also, smolagents already include the tool call in the prompt,
+                # so adding `toolConfig` could cause conflicts. We remove it to avoid issues.
+                self.bedrock_api_kwargs.pop("toolConfig")
+            request_payload.update(**self.bedrock_api_kwargs)
```

Comment:
This could be moved to your specific `_prepare_completion_kwargs` implementation.

Reply:
Agree!